File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/realm/models/Variable.java
package ch.rmy.android.http_shortcuts.realm.models;

import android.content.Context;

import ch.rmy.android.http_shortcuts.R;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.Required;

public class Variable extends RealmObject implements HasId {

    public static final String FIELD_KEY = "key";

    public static final String TYPE_CONSTANT = "constant";
    public static final String TYPE_TEXT = "text";
    public static final String TYPE_NUMBER = "number";
    public static final String TYPE_PASSWORD = "password";
    public static final String TYPE_SELECT = "select";
    public static final String TYPE_TOGGLE = "toggle";
    public static final String TYPE_COLOR = "color";

    public static final String[] TYPE_OPTIONS = {
            TYPE_CONSTANT,
            TYPE_TEXT,
            TYPE_NUMBER,
            TYPE_PASSWORD,
            TYPE_COLOR,
            TYPE_SELECT,
            TYPE_TOGGLE,
    };
    public static final int[] TYPE_RESOURCES = {
            R.string.variable_type_constant,
            R.string.variable_type_text,
            R.string.variable_type_number,
            R.string.variable_type_password,
            R.string.variable_type_color,
            R.string.variable_type_select,
            R.string.variable_type_toggle
    };

    private static final int FLAG_SHARE_TEXT = 0x1;

    @PrimaryKey
    private long id;

    @Required
    private String key;
    @Required
    private String type;

    private String value;
    private RealmList<Option> options;

    private boolean rememberValue;
    private boolean urlEncode;
    private boolean jsonEncode;

    private int flags;

    @Required
    private String title;

    public static Variable createNew() {
        Variable variable = new Variable();
        variable.setKey("");
        variable.setType(TYPE_CONSTANT);
        variable.setValue("");
        variable.setTitle("");
        variable.setOptions(new RealmList<Option>());
        return variable;
    }

    @Override
    public boolean isNew() {
        return id == 0;
    }

    @Override
    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    public RealmList<Option> getOptions() {
        return options;
    }

    public void setOptions(RealmList<Option> options) {
        this.options = options;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isRememberValue() {
        return rememberValue;
    }

    public void setRememberValue(boolean rememberValue) {
        this.rememberValue = rememberValue;
    }

    public boolean isUrlEncode() {
        return urlEncode;
    }

    public void setUrlEncode(boolean urlEncode) {
        this.urlEncode = urlEncode;
    }

    public boolean isJsonEncode() {
        return jsonEncode;
    }

    public void setJsonEncode(boolean jsonEncode) {
        this.jsonEncode = jsonEncode;
    }

    public int getFlags() {
        return flags;
    }

    public void setFlags(int flags) {
        this.flags = flags;
    }

    public boolean isShareText() {
        return (getFlags() & FLAG_SHARE_TEXT) != 0;
    }

    public static String[] getTypeOptions(Context context) {
        String[] typeStrings = new String[Variable.TYPE_OPTIONS.length];
        for (int i = 0; i < Variable.TYPE_OPTIONS.length; i++) {
            typeStrings[i] = context.getString(Variable.TYPE_RESOURCES[i]);
        }
        return typeStrings;
    }

    public boolean isResetAfterUse() {
        return !isRememberValue() && (TYPE_TEXT.equals(getType()) || TYPE_NUMBER.equals(getType()) || TYPE_PASSWORD.equals(getType()) || TYPE_COLOR.equals(getType()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Variable variable = (Variable) o;

        if (isRememberValue() != variable.isRememberValue()) return false;
        if (isUrlEncode() != variable.isUrlEncode()) return false;
        if (isJsonEncode() != variable.isJsonEncode()) return false;
        if (!getKey().equals(variable.getKey())) return false;
        if (!getType().equals(variable.getType())) return false;
        if (!getValue().equals(variable.getValue())) return false;
        if (!getOptions().equals(variable.getOptions())) return false;
        return getTitle().equals(variable.getTitle());
    }

    @Override
    public int hashCode() {
        int result = getKey().hashCode();
        result = 31 * result + getType().hashCode();
        result = 31 * result + getValue().hashCode();
        result = 31 * result + getOptions().hashCode();
        result = 31 * result + (isRememberValue() ? 1 : 0);
        result = 31 * result + (isUrlEncode() ? 1 : 0);
        result = 31 * result + (isJsonEncode() ? 1 : 0);
        result = 31 * result + getTitle().hashCode();
        return result;
    }
}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/realm/models/Variable.kt
package ch.rmy.android.http_shortcuts.realm.models

import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.utils.GsonUtil
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import io.realm.annotations.Required


open class Variable : RealmObject(), HasId {

    @PrimaryKey
    override var id: Long = 0

    @Required
    var key: String = ""
    @Required
    var type: String = ""

    var value: String? = null
    var options: RealmList<Option>? = null

    var rememberValue: Boolean = false
    var urlEncode: Boolean = false
    var jsonEncode: Boolean = false

    var data: String? = null

    var flags: Int = 0

    @Required
    var title: String = ""

    override val isNew: Boolean
        get() = id == 0L

    var isShareText: Boolean
        get() = flags and FLAG_SHARE_TEXT != 0
        set(shareText) = if (shareText) {
            flags = flags or FLAG_SHARE_TEXT
        } else {
            flags = flags and FLAG_SHARE_TEXT.inv()
        }

    fun isResetAfterUse(): Boolean =
            !rememberValue && type in listOf(TYPE_TEXT, TYPE_NUMBER, TYPE_PASSWORD, TYPE_COLOR, TYPE_SLIDER)

    fun isSameAs(other: Variable): Boolean {
        if (other.key != key ||
                other.type != type ||
                other.value != value ||
                other.title != title ||
                other.options!!.size != options!!.size
                ) {
            return false
        }
        if (other.options!!.indices.any { !options!![it]!!.isSameAs(other.options!![it]!!) }) {
            return false
        }
        return true
    }

    var dataForType: Map<String, String?>
        get() = GsonUtil.fromJsonObject<Map<String, String?>>(data)[type]?.toMap() ?: emptyMap()
        set(value) {
            val dataMap = GsonUtil.fromJsonObject<Map<String, String?>>(data).toMutableMap()
            dataMap[type] = value
            data = GsonUtil.toJson(dataMap)
        }

    companion object {

        const val FIELD_KEY = "key"

        const val TYPE_CONSTANT = "constant"
        const val TYPE_TEXT = "text"
        const val TYPE_NUMBER = "number"
        const val TYPE_PASSWORD = "password"
        const val TYPE_SELECT = "select"
        const val TYPE_TOGGLE = "toggle"
        const val TYPE_COLOR = "color"
        const val TYPE_DATE = "date"
        const val TYPE_TIME = "time"
        const val TYPE_SLIDER = "slider"

        val TYPE_OPTIONS = arrayOf(TYPE_CONSTANT, TYPE_TEXT, TYPE_NUMBER, TYPE_PASSWORD, TYPE_DATE, TYPE_TIME, TYPE_COLOR, TYPE_SELECT, TYPE_TOGGLE, TYPE_SLIDER)
        val TYPE_RESOURCES = intArrayOf(R.string.variable_type_constant, R.string.variable_type_text, R.string.variable_type_number, R.string.variable_type_password, R.string.variable_type_date, R.string.variable_type_time, R.string.variable_type_color, R.string.variable_type_select, R.string.variable_type_toggle, R.string.variable_type_slider)

        private const val FLAG_SHARE_TEXT = 0x1

        fun createNew(): Variable {
            val variable = Variable()
            variable.key = ""
            variable.type = TYPE_CONSTANT
            variable.value = ""
            variable.title = ""
            variable.options = RealmList<Option>()
            return variable
        }

    }

}
